btf/types.go

```diff
@@ -0,0 +1,615 @@
+				Name:   nameByOffset(member.NameOff),
+				Type:   typeByID(member.Type),
+				Offset: member.Offset,
+			})
+		}
+		return members
+	}
+
+	typeByID = func(id TypeID) Type {
+		typ, ok := byID[id]
+		if ok {
+			return typ
+		}
+
+		if int(id-1) >= len(rawTypes) {
+			panic(errors.Errorf("invalid type id %d", id))
+		}
+
+		var (
+			raw  = rawTypes[int(id-1)]
+			name = nameByOffset(raw.NameOff)
+		)
+
+		if visited.contains(id) {
+			panic(errors.Errorf("circular type %v %d: %v", name, id, visited))
+		}
+
+		visited.push(id, name)
+		defer visited.pop()
+
+		if len(visited) > maxTypeDepth {
+			panic(errors.Errorf("exceeded maximum depth: %v", visited))
+		}
+
+		switch raw.Kind() {
+		case kindInt:
+			typ = &Int{id, name, raw.Size()}
+
+		case kindPointer:
+			// structs and unions may have members which are pointers
+			// to themselves.
+			//    struct foo { struct foo *next; }
+			//    KIND_STRUCT -> KIND_PTR -> KIND_STRUCT
+			//    KIND_PTR -> KIND_STRUCT -> KIND_PTR
+			// Collect them here, and resolve them later.
+			ptr := &Pointer{TypeID: id}
+			pointers[raw.Type()] = ptr
+			typ = ptr
+
+		case kindArray:
+			arr := raw.data.(*btfArray)
+
+			// IndexType is unused according to btf.rst.
+			// Try to resolve it, but don't make it available
+			// right now.
+			_ = typeByID(arr.IndexType)
+
+			typ = &Array{
+				TypeID: id,
+				Type:   typeByID(arr.Type),
+				Nelems: arr.Nelems,
+			}
+
+		case kindStruct:
+			typ = &Struct{id, name, raw.Size(), convertMembers(raw.data.([]btfMember))}
+
+		case kindUnion:
+			typ = &Union{id, name, raw.Size(), convertMembers(raw.data.([]btfMember))}
+
+		case kindEnum:
+			typ = &Enum{id, name}
+
+		case kindForward:
+			typ = &Fwd{id, name}
+
+		case kindTypedef:
+			typ = &Typedef{id, name, typeByID(raw.Type())}
+
+		case kindVolatile:
+			typ = &Volatile{id, typeByID(raw.Type())}
+
+		case kindConst:
+			typ = &Const{id, typeByID(raw.Type())}
+
+		case kindRestrict:
+			typ = &Restrict{id, typeByID(raw.Type())}
+
+		case kindFunc:
+			// TODO: Check that Type is a FuncProto?
+			typ = &Func{id, name, typeByID(raw.Type())}
+
+		case kindFuncProto:
+			typ = &FuncProto{id, typeByID(raw.Type())}
+
+		case kindVar:
+			typ = &Var{id, name, typeByID(raw.Type())}
+
+		case kindDatasec:
+			typ = &Datasec{id, name, raw.SizeType}
+
+		default:
+			panic(errors.Errorf("type id %d: unknown kind: %v", id, raw.Kind()))
+		}
+
+		byID[id] = typ
+		return typ
+	}
+
+	types := make(map[string][]Type, len(rawTypes))
+	for i := range rawTypes {
+		id := TypeID(i + 1)
+		typ := typeByID(id)
+
+		if namer, ok := typ.(namer); ok && namer.name() != "" {
+			name := namer.name()
+			types[name] = append(types[name], typ)
+		}
+	}
+
+	for id, pointer := range pointers {
+		target, ok := byID[id]
+		if !ok {
+			return nil, errors.Errorf("can't resolve pointer to type id %d", id)
+		}
+
+		pointer.Target = target
+	}
+
+	return types, nil
+}
+
+type typeStack []typeStackEntry
```

Comment:
I think it would work, at the expense of making inflateTypes even longer. There would need to be something like:

```
 defer func() { delete(stack, typeStackEntry{} }()
```

It also means that we lose the ordered debug output, which currently shows the chain of types that led to the circular reference.

Reply:
Wouldn't it be the same as the existing `defer`, just `defer delete(stack, entry)`? Fair point about debugging though.

Reply:
This is now gone.